src/awkward/_v2/cpp-headers/LayoutBuilder.h

```diff
@@ -0,0 +1,534 @@
+// BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+#ifndef AWKWARD_LAYOUTBUILDER_H_
+#define AWKWARD_LAYOUTBUILDER_H_
+
+#include "GrowableBuffer.h"
+#include "utils.h"
+
+#include <stdexcept>
+#include <cassert>
+#include <tuple>
+
+namespace awkward {
+
+  template<const char *str>
+  struct field_name {
+    const char* value = str;
+  };
+
+  template<class field_name, class BUILDER>
+  struct Record {
+    const char* field() {
+        return field_.value;
+    }
+
+    std::string
+    form() {
+      return builder.form();
+    }
+
+    field_name field_;
+    BUILDER builder;
+  };
+
+  template <size_t INDEX>
+  struct visit_impl
+  {
+      template <typename RECORD, typename FUNCTION>
+      static void visit(RECORD& contents, size_t index, FUNCTION fun)
+      {
+          if (index == INDEX - 1) fun(std::get<INDEX - 1>(contents));
+          else visit_impl<INDEX - 1>::visit(contents, index, fun);
+      }
+  };
+
+  template <>
+  struct visit_impl<0>
+  {
+      template <typename RECORD, typename FUNCTION>
+      static void visit(RECORD& contents, size_t index, FUNCTION fun) { assert(false); }
+  };
+
+  template <typename FUNCTION, typename... RECORDs>
+  void
+  visit_at(std::tuple<RECORDs...> const& contents, size_t index, FUNCTION fun)
+  {
+      visit_impl<sizeof...(RECORDs)>::visit(contents, index, fun);
+  }
+
+  template <typename FUNCTION, typename... RECORDs>
+  void
+  visit_at(std::tuple<RECORDs...>& contents, size_t index, FUNCTION fun)
+  {
+      visit_impl<sizeof...(RECORDs)>::visit(contents, index, fun);
+  }
+
+  template <unsigned ID, unsigned INITIAL, typename PRIMITIVE>
+  class NumpyLayoutBuilder {
+  public:
+    NumpyLayoutBuilder()
+        : data_(awkward::GrowableBuffer<PRIMITIVE>(INITIAL)) { }
+
+    size_t
+    length() const {
+      return data_.length();
+    }
+
+    void
+    clear() {
+      data_.clear();
+    }
+
+    void
+    append(PRIMITIVE x) {
+      data_.append(x);
+    }
+
+    void
+    append(PRIMITIVE* ptr, size_t size) {
+      data_.append(ptr, size);
+    }
+
+    void
+    to_buffers(PRIMITIVE* ptr) const {
+      data_.concatenate(ptr);
+    }
+
+    std::string
+    form() {
+      std::stringstream form_key;
+      form_key << "node" << id_;
+      if (std::is_arithmetic<PRIMITIVE>::value) {
+        return "{ \"class\": \"NumpyArray\", \"primitive\": \""
+                  + type_to_name<PRIMITIVE>() + "\", " + "\"form_key\": \"" + form_key.str() + "\" }";
+      } else if (is_specialization<PRIMITIVE, std::complex>::value) {
+        return "{ \"class\": \"NumpyArray\", \"primitive\": \"complex128\", \"form_key\": \""
+                  + form_key.str() + "\" }";
+      }
+      return "unsupported type";
```

Comment:
perhaps, it would be better to add more information on which type is not supported yet? ` typeid(PRIMITIVE).name()`?